### PR TITLE
Js doc support

### PR DESCRIPTION
**Pull request by RyanCavanaugh**

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -166,9 +169,50 @@ namespace ts {
...
+            Debug.assert(isAmdExportAssignment(<BinaryExpression>node));
+
+            let binaryExpr = <BinaryExpression>node;
+            let propAccess = <PropertyAccessExpression>(binaryExpr.left);
+            return propAccess.name.text;
+        }
+
+        function getAnonymousModuleName(node: Node) {
+            let parent = node.parent;
+            while (parent && parent.kind !== SyntaxKind.SourceFile) {
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: There is a helper for this

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -166,9 +169,50 @@ namespace ts {
...
+                parent = parent.parent;
+            }
+
+            let sourceFileName = removeFileExtension((<SourceFile>parent).fileName);
+            return sourceFileName && ('"' + sourceFileName + '"');
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: This should always be truthy

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -835,9 +886,13 @@ namespace ts {
 
         function bindWorker(node: Node) {
             switch (node.kind) {
+                /* Strict mode checks */
                 case SyntaxKind.Identifier:
                     return checkStrictModeIdentifier(<Identifier>node);
                 case SyntaxKind.BinaryExpression:
+                    if(isAmdExportAssignment(<BinaryExpression>node)) {
+                        return bindAmdExportAssignment(<BinaryExpression>node);
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Don't return here

File: src/compiler/binder.ts [hunk 4]
```diff
@@ -852,6 +907,12 @@ namespace ts {
                 case SyntaxKind.WithStatement:
                     return checkStrictModeWithStatement(<WithStatement>node);
 
+                case SyntaxKind.CallExpression:
+                    if (isDefineCall(<CallExpression>node)) {
+                        return bindDefineCall(<CallExpression>node);
+                    }
+                    break;
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: return not break

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -969,6 +969,39 @@ namespace ts {
...
+        return isInJavaScriptFile(expression) && isCalledToNamedFunction(expression, 'define');
+    }
+
+    export function isAmdRequireCall(expression: CallExpression): boolean {
+        return isInJavaScriptFile(expression) && isCalledToNamedFunction(expression, 'require') && expression.arguments.length > 1;
+    }
+
+    export function isAmdExportAssignment(expression: Node): boolean;
+    export function isAmdExportAssignment(expression: BinaryExpression): boolean {
+        return (expression.kind === SyntaxKind.BinaryExpression) &&
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Check that it's a JS file

Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: And that the parent is an ExpressionStatement

File: src/compiler/utilities.ts [hunk 6]
```diff
@@ -969,6 +969,39 @@ namespace ts {
...
+    }
+
+    function isCalledToNamedFunction(expression: CallExpression, name: string) {
+        return expression.expression.kind === SyntaxKind.Identifier &&
+                (<Identifier>expression.expression).text === name;
+    }
+
+    export function isDefineCall(expression: CallExpression): boolean {
+        // In .js files, calls to the identifier 'define' are treated specially
+        // Walk up to the source file parent
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Stale comment

File: src/compiler/utilities.ts [hunk 7]
```diff
@@ -969,6 +969,39 @@ namespace ts {
...
+        return isInJavaScriptFile(expression) && isCalledToNamedFunction(expression, 'define');
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Only do this when the compiler settings specify an appropriate module system

File: src/compiler/utilities.ts [hunk 8]
```diff
@@ -969,6 +969,39 @@ namespace ts {
...
+    }
+
+    export function isAmdRequireCall(expression: CallExpression): boolean {
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Give an example

File: src/compiler/binder.ts [hunk 9]
```diff
@@ -966,6 +1030,26 @@ namespace ts {
             }
         }
 
+        function bindAmdExportAssignment(node: BinaryExpression) {
+            let symbol = declareSymbolAndAddToSymbolTableWorker(node, SymbolFlags.ExportValue | SymbolFlags.Property, SymbolFlags.None);
+            symbol.isAmdExportAssignment = true;
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Maybe we can do this syntactically

File: src/compiler/binder.ts [hunk 10]
```diff
@@ -966,6 +1030,26 @@ namespace ts {
...
+        function bindAmdExportAssignment(node: BinaryExpression) {
+            let symbol = declareSymbolAndAddToSymbolTableWorker(node, SymbolFlags.ExportValue | SymbolFlags.Property, SymbolFlags.None);
+            symbol.isAmdExportAssignment = true;
+            if (symbol.exportSymbol) {
+                symbol.exportSymbol.isAmdExportAssignment = true;
+            }
+        }
+
+        function bindDefineCall(node: CallExpression) {
+            let symbol = declareSymbolAndAddToSymbolTableWorker(node, SymbolFlags.ValueModule | SymbolFlags.ExportValue, SymbolFlags.None);
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Try removing ExportValue?

File: src/compiler/binder.ts [hunk 11]
```diff
@@ -966,6 +1030,26 @@ namespace ts {
...
+            symbol.isDefineModule = true;
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Same as above, try to do this syntactically

File: src/compiler/binder.ts [hunk 12]
```diff
@@ -966,6 +1030,26 @@ namespace ts {
...
+
+            // If this was a file-level module, hook up the file symbol to this module
+            let parent = node.parent;
```
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: This should only happen if it's anonymous

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -1369,7 +1373,8 @@ namespace ts {
 
         function hasExternalModuleSymbol(declaration: Node) {
             return (declaration.kind === SyntaxKind.ModuleDeclaration && (<ModuleDeclaration>declaration).name.kind === SyntaxKind.StringLiteral) ||
-                (declaration.kind === SyntaxKind.SourceFile && isExternalModule(<SourceFile>declaration));
+                (declaration.kind === SyntaxKind.SourceFile && isExternalModule(<SourceFile>declaration)) ||
+                (declaration.kind === SyntaxKind.CallExpression && isDefineCall(<CallExpression>declaration));
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Make `isDefineCall` take a `Node`

Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Set `externalModuleIndicator` in the binder

File: src/compiler/checker.ts [hunk 14]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
             return links.type;
         }
 
+        function resolveAmdExportAssignment(symbol: Symbol): Type {
+            let seenTypes: Type[] = [];
+
+            for (var i = 0; i < symbol.declarations.length; i++) {
+                let decl = symbol.declarations[i];
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Make sure we have tests for multiple declarations

File: src/compiler/checker.ts [hunk 15]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            for (var i = 0; i < symbol.declarations.length; i++) {
+                let decl = symbol.declarations[i];
+                Debug.assert(isAmdExportAssignment(decl));
+                seenTypes.push(getTypeOfExpression((<BinaryExpression>symbol.declarations[i]).right));
+            }
+
+            return getUnionType(seenTypes);
+        }
+
+        function resolveDefineModule(symbol: Symbol): Type {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Add comment explaining what this does

File: src/compiler/checker.ts [hunk 16]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            // Can't meaningfully define the same module more than once
+            if(symbol.declarations.length !== 1) {
+                return unknownType;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Should just Debug.fail here

File: src/compiler/checker.ts [hunk 17]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            }
+
+            // If the invocation of 'define' has zero args ('define()')... we
+            // shouldn't really be here, but it's meaninguless
+            let callExpr = <CallExpression>symbol.declarations[0];
+            if(callExpr.arguments.length === 0) {
+                return unknownType;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Debug.fail

File: src/compiler/checker.ts [hunk 18]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            }
+
+            // If the last arg isn't a function expr, just resolve
+            // its type and define that as the shape of the module
+            let lastArg = <FunctionExpression>callExpr.arguments[callExpr.arguments.length - 1];
+            if(lastArg.kind !== SyntaxKind.FunctionExpression) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Use isFunctionLike / etc to allow arrow functions here; add test

File: src/compiler/checker.ts [hunk 19]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+                return bodyReturnType;
+            }
+
+            // Collect assignments in the body to 'module.exports' or 'exports.propName' or 'module.exports.propName'
+            let assignedModuleType: Type = undefined;
+            let exportAssignedProperties: Symbol[] = [];
+            let exportAssignedPropTable: SymbolTable = {};
+            
+            let cjsModuleType = getExportedTypeFromNamespace('CommonJS', 'Module');
+            Debug.assert(cjsModuleType !== undefined && !isTypeAny(cjsModuleType));
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Return unknownType or something here

File: src/compiler/checker.ts [hunk 20]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            let cjsExportsType = getExportedTypeFromNamespace('CommonJS', 'Exports');
+            Debug.assert(cjsExportsType !== undefined && !isTypeAny(cjsExportsType));
+            function traverse(node: Node) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Move to bottom of enclosing function

File: src/compiler/checker.ts [hunk 21]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+                if(node.kind === SyntaxKind.BinaryExpression &&
+                   (<BinaryExpression>node).operatorToken.kind === SyntaxKind.EqualsToken) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Give examples

File: src/compiler/checker.ts [hunk 22]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+                    let lhs = (<BinaryExpression>node).left;
+                    if(lhs.kind === SyntaxKind.PropertyAccessExpression) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Add mutually recursive dependencies here

File: src/compiler/checker.ts [hunk 23]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            traverse(lastArg.body);
+
+            // Someone assigned to 'module.exports', so ignore other assignments
+            if (assignedModuleType) {
+                return assignedModuleType;
+            }
+
+            // Otherwise return the collected property assignments
+            if (exportAssignedProperties.length > 0) {
+                return createAnonymousType(lastArg.symbol, exportAssignedPropTable, [], [], undefined, undefined);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Named args

File: src/compiler/checker.ts [hunk 24]
```diff
@@ -2656,7 +2723,122 @@ namespace ts {
...
+            }
+
+            // Empty module?
+            return emptyObjectType;
+        }
+
+        function getTypeOfDefineModule(symbol: Symbol): Type {
+            let links = getSymbolLinks(symbol);
+            if (!links.type) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Set links.type to the recursive guard thingy

File: src/compiler/checker.ts [hunk 25]
```diff
@@ -9077,6 +9459,15 @@ namespace ts {
                     return anyType;
                 }
             }
+
+            let exprType = getTypeOfExpression(node.expression);
+            if((exprType === getExportedTypeFromNamespace('CommonJS', 'Require')) ||
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Cache this

File: src/compiler/checker.ts [hunk 26]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
                 pos < signature.parameters.length ? getTypeOfSymbol(signature.parameters[pos]) : anyType;
         }
 
+        // A function expression is a CommonJS Wrapper function if it has
+        // 1 parameter named 'require', two parameters named 'require' and 'exports',
+        // or 3 parameters named 'require', 'exports', and 'module' in that exact order
+        function isCommonJsWrapper(expression: FunctionExpression): boolean {
+            // All fall-throughs in this switch are intentional
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: // fall-through at each site

File: src/compiler/checker.ts [hunk 27]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
                 pos < signature.parameters.length ? getTypeOfSymbol(signature.parameters[pos]) : anyType;
         }
 
+        // A function expression is a CommonJS Wrapper function if it has
+        // 1 parameter named 'require', two parameters named 'require' and 'exports',
+        // or 3 parameters named 'require', 'exports', and 'module' in that exact order
+        function isCommonJsWrapper(expression: FunctionExpression): boolean {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Take `FunctionExpression|ArrowFunction`

File: src/compiler/checker.ts [hunk 28]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
...
+                    if ((<Identifier>expression.parameters[0].name).text !== 'require') {
+                        return false;
+                    }
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        function assignDefineOrRequireCallParameterTypes(funcExpr: FunctionExpression) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Take arrow functions as well

File: src/compiler/checker.ts [hunk 29]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
...
+                    case 1:
+                        getSymbolLinks(funcExpr.parameters[0].symbol).type = getExportedTypeFromNamespace('CommonJS', 'Require');
+                }
+
+                return;
+            }
+
+            // Example: define(['my', 'dependency', 'array'], function(m, d, a) { ... } )
+            if (hasDependencyArray) {
+                for (let i = 0; i < funcExpr.parameters.length; i++) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Comment that we might be ignoring some module names on purpose

File: src/compiler/checker.ts [hunk 30]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
...
+                    if (moduleNames.elements.length === i) {
+                        // We exhausted the list of modules. TODO: Issue an error; this is bad.
+                        break;
+                    }
+
+                    let links = getSymbolLinks(funcExpr.parameters[i].symbol);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Move this inside the `if`

File: src/compiler/checker.ts [hunk 31]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
...
+
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Short-circuit to `any` if we have some other kind of parameter here

File: src/compiler/checker.ts [hunk 32]
```diff
@@ -9102,6 +9493,82 @@ namespace ts {
...
+                            links.type = getExportedTypeFromNamespace('CommonJS', 'Require');
+                        }
+                        else if (unquotedName === 'exports') {
+                            links.type = getExportedTypeFromNamespace('CommonJS', 'Exports');
+                        }
+                        else if (unquotedName === 'module') {
+                            links.type = getExportedTypeFromNamespace('CommonJS', 'Module');
+                        }
+                        else {
+                            links.type = resolveExternalModuleTypeByLiteral(<StringLiteral>moduleNames.elements[i]);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Already have `moduleName` in scope

File: src/compiler/checker.ts [hunk 33]
```diff
@@ -9404,6 +9886,11 @@ namespace ts {
                 }
             }
 
+            // Handle 'define' calls
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: `isDefineCall` should require at least one arg

File: src/compiler/checker.ts [hunk 34]
```diff
@@ -13804,6 +14291,11 @@ namespace ts {
                     }
 
                     switch (location.kind) {
+                        case SyntaxKind.CallExpression:
+                            if(isDefineCall(<CallExpression>location)) {
+                                copySymbols(getSymbolOfNode(location).exports, meaning & SymbolFlags.ModuleMember);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: This is probably totally wrong

File: src/compiler/parser.ts [hunk 35]
```diff
@@ -5233,6 +5235,7 @@ namespace ts {
                     || node.kind === SyntaxKind.ImportDeclaration
                     || node.kind === SyntaxKind.ExportAssignment
                     || node.kind === SyntaxKind.ExportDeclaration
+                    || node.kind === SyntaxKind.ExpressionStatement && isAmdExportAssignment((<ExpressionStatement>node).expression)
```
Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: Also do this for `require` calls

File: src/compiler/utilities.ts [hunk 36]
```diff
@@ -969,6 +969,39 @@ namespace ts {
         return node.kind === SyntaxKind.ImportEqualsDeclaration && (<ImportEqualsDeclaration>node).moduleReference.kind !== SyntaxKind.ExternalModuleReference;
     }
 
+    function isInJavaScriptFile(node: Node): boolean {
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Remove, check node.flags for JavaScript

File: src/lib/core.d.ts [hunk 37]
```diff
@@ -1182,3 +1182,37 @@ interface PromiseLike<T> {
...
+        (moduleName: string, dependencies: string[], callback: (...resolvedModules: any[]) => any): void;
+        (func: (require: Require, module: { exports: any }, exports: any) => any): void;
+        (func: (require: Require, module: { exports: any }, exports: any) => any): void;
+        (object: {}): void;
+    }
+
+    export var define: Define;
+}
+
+declare namespace CommonJS {
```
Comment (RyanCavanaugh) [inline src/lib/core.d.ts]: Make sure this doesn't conflict with anything on DefinitelyTyped

File: src/lib/core.d.ts [hunk 38]
```diff
@@ -1182,3 +1182,37 @@ interface PromiseLike<T> {
...
+        (moduleList: string[], callback: (...resolvedModules: any[]) => void): any;
+    }
+
+    export var require: Require;
+
+    interface Define {
+        (dependencies: string[], callback: (...resolvedModules: any[]) => any): void;
+        (moduleName: string, dependencies: string[], callback: (...resolvedModules: any[]) => any): void;
+        (func: (require: Require, module: { exports: any }, exports: any) => any): void;
+        (func: (require: Require, module: { exports: any }, exports: any) => any): void;
```
Comment (RyanCavanaugh) [inline src/lib/core.d.ts]: Remove

File: src/lib/core.d.ts [hunk 39]
```diff
@@ -1182,3 +1182,37 @@ interface PromiseLike<T> {
...
+        (object: {}): void;
```
Comment (RyanCavanaugh) [inline src/lib/core.d.ts]: (moduleName: string, object: {})

**Comment by zhengbli** [comment]
What does this PR do? Is it related to #2646?

File: src/compiler/binder.ts (hunk 1 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: :heavy_check_mark:

File: src/compiler/binder.ts (hunk 2 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: :heavy_check_mark:

File: src/compiler/binder.ts (hunk 3 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: :heavy_check_mark:

File: src/compiler/binder.ts (hunk 4 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: :heavy_check_mark:

File: src/compiler/utilities.ts (hunk 5 above)
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: :heavy_check_mark:

File: src/compiler/utilities.ts (hunk 6 above)
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: :heavy_check_mark:

File: src/compiler/utilities.ts (hunk 7 above)
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Address this aspect in calling code

File: src/compiler/utilities.ts (hunk 8 above)
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: :heavy_check_mark:

File: src/compiler/checker.ts [hunk 40]
```diff
@@ -4332,9 +4669,37 @@ namespace ts {
                 case SyntaxKind.QualifiedName:
                     let symbol = getSymbolAtLocation(node);
                     return symbol && getDeclaredTypeOfSymbol(symbol);
-                default:
+                case SyntaxKind.JSDocAllType:
+                    return anyType;
+                case SyntaxKind.JSDocUnknownType:
                     return unknownType;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This should be `anyType`

File: src/compiler/binder.ts (hunk 9 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Yep!

File: src/compiler/binder.ts (hunk 10 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Yep

File: src/compiler/binder.ts (hunk 11 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: :heavy_check_mark:

File: src/compiler/binder.ts (hunk 12 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 13 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 15 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 17 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 18 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 19 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 20 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 21 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 23 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 24 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 25 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 26 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 27 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 28 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 29 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 30 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 31 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 33 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 34 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/parser.ts (hunk 35 above)
Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: :heavy_check_mark:

File: src/compiler/utilities.ts (hunk 36 above)
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: :heavy_check_mark:

File: src/lib/core.d.ts (hunk 38 above)
Comment (RyanCavanaugh) [inline src/lib/core.d.ts]: :heavy_check_mark:

File: src/lib/core.d.ts (hunk 37 above)
Comment (RyanCavanaugh) [inline src/lib/core.d.ts]: :heavy_check_mark:

File: src/lib/core.d.ts (hunk 39 above)
Comment (RyanCavanaugh) [inline src/lib/core.d.ts]: :heavy_check_mark:

File: src/compiler/checker.ts [hunk 41]
```diff
@@ -965,7 +970,15 @@ namespace ts {
...
-            if (!moduleName) return;
+            if(moduleName.indexOf('!') >= 0) {
+                moduleName = moduleName.substr(0, moduleName.indexOf('!'));
+            }
+
+            if (!moduleName) {
+                return;
+            }
+
+            debugger;
```
Comment (billti) [inline src/compiler/checker.ts]: This is *really* annoying while running under the debugger :smile:

**Comment by mhegazy** [comment]
do we still need this PR, or is branch https://github.com/Microsoft/TypeScript/tree/JSDocAndJSModules sufficient?

**Comment by RyanCavanaugh** [comment]
Tracking at #5044

**Pull request closed without merging**